Set guest vagrant provision with python, pip & Flask.

DudiChen opened the pull request:
Guest machine initiates with Flask Hello World via Host machine @ localhost:5000

DudiChen committed:
Set guest vagrant provision with python, pip & Flask. Guest initiated with Flask Hello World via Host localhost:5000

DudiChen committed:
1. Added Vagrantfile comments for re-spinning of existing guest image 2. Enabled auto init of Flask server on guest reboot.

DudiChen committed:
changed - char to + char in line 'chmod -x $FLASK_APP'

DudiChen committed:
1. Upgraded guest to Ubuntu-18.0.4(bionic64)
  Updated Vagrantfile & setup.sh with upgrade requirements
2. Changed name of flask_init.txt to flask_init.py
3. Added runFlask.sh & killFlask.sh

DudiChen committed:
setup.sh: Fix: replaced "flaskOnBoot.sh" to runFlask.sh"

simzacks commented on Vagrantfile:
```diff
@@ -1,7 +1,11 @@
 Vagrant.configure("2") do |config|
   config.vm.box = "ubuntu/xenial64"
+  config.vm.provision :shell, path: "setup.sh", :privileged => false 
+  #config.vm.network :public_network, ip: "192.168.50.6"
```
please remove unnecessary comments.

DudiChen replied:
Done. Please see latest commit [61f0cee].

lmilbaum replied:
@DudiChen A shorter convention of accepting a comment review is to use one of the following:
ack (for acknowledge) or +1
Both terms are widely used by the open source community

simzacks commented on Vagrantfile:
```diff
@@ -1,7 +1,11 @@
+  config.vm.provision :shell, path: "setup.sh", :privileged => false 
+  #config.vm.network :public_network, ip: "192.168.50.6"
+  config.vm.network :private_network, ip: "192.168.50.6"
+  config.vm.network :forwarded_port, guest: 5000, host: 5000
```
Is there a reason you  you choose this IP address?

DudiChen replied:
No. Arbitrary IP in the private network address scope to enable comm between host and guest,
Found it unnecessary as the forwarded_port config is not dependent on a private network IP address.
Removed this line in latest commit [61f0cee].

simzacks commented on flask_init.txt:
```diff
@@ -0,0 +1,12 @@
+@app.route("/")
+def index():
+	return make_response('Hello World!', 200)
+    #return Response("Hello World!"), 200
```
please remove unnecessary comments

DudiChen replied:
Done. Please see latest commit [61f0cee].

simzacks commented on flask_init.txt:
```diff
@@ -0,0 +1,12 @@
+
+@app.route("/")
+def index():
+	return make_response('Hello World!', 200)
```
This should return an HTML page, not just text.

DudiChen replied:
Done. Please see latest commit [61f0cee].

simzacks commented on flask_init.txt:
```diff
@@ -0,0 +1,12 @@
+from flask import Flask, make_response
+# from flask import Flask, Response
```
please remove unnecessary comments

DudiChen replied:
Done. Please see latest commit [61f0cee].

simzacks commented on setup.sh:
```diff
@@ -0,0 +1,66 @@
+    [ $? -ne 0 ] && exit 1
+
+    sudo DEBIAN_FRONTEND=noninteractive apt-get -yq install python3-dev python3-pip  >> $OUTFILE 2>&1 
+    [ $? -ne 0 ] && exit 1
```
why are you trying to get pip manually and then installing it with package manager (apt) afterwards?
If you are -yq options, why are you also using DEBIAN_FRONTEND=noninteractive?

DudiChen replied:
Noted. Upgraded OS version and removed these lines from setup.sh accordingly.
Please see latest commit [61f0cee].

DudiChen committed:
Modified flask_init.py to return HTML
Added static and templates dir with HTML content

DudiChen committed:
[Vagrantfile] Removed unneccessary private_network configuration

DudiChen committed:
[Vagrantfile] Removed comments

lmilbaum commented on runFlask.sh:
```diff
@@ -0,0 +1,14 @@
+FLASK_DIR=$HOME/FlaskApp
+
+cd $FLASK_DIR
+if [ -e venv/bin/activate ]; then
```
Why a virtualenv was required?

simzacks replied:
IOW, the entire vagrant is a virtualenv, this puts a virtualenv into a virtualenv.

DudiChen replied:
In short, a best practice to isolate a Python project with its own required libraries to ensure all dependencies.
It's reasonable that this won't necessarily be an issue in our case, I just applied it per Flask documentation recommendation.

Taken from Flask Documentation
https://flask.palletsprojects.com/en/1.1.x/installation/

"The more Python projects you have, the more likely it is that you need to work with different versions of Python libraries, or even Python itself. Newer versions of libraries for one project can break compatibility in another project.

Virtual environments are independent groups of Python libraries, one for each project. Packages installed for one project will not affect other projects or the operating system’s packages."

lmilbaum replied:
@DudiChen The documentation is correct, but, not relevant to case of spining a DEV environment with vagrant. It is by itself an isolated environment.

lmilbaum commented on setup.sh:
```diff
@@ -0,0 +1,74 @@
+    echo "==> Running Provision Script: $0:"
+    echo " "
+    echo "$((step++)). Installing & Activating Python-3 Virtual Environment..."
+    sudo apt-get install -yq python3-venv >> $OUTFILE 2>&1 
```
why sudo is required?

DudiChen replied:
In case it would require to write system directories.
In this instance I admit I did not try it without sudo.
I will test it and omit it if redundant.

DudiChen replied:
Indeed found sudo to be mandatory for all required apt-get and pip operations for permissions purposes of certain paths.

lmilbaum replied:
It is mandatory when you provision with ":privileged => false".  Remove it and you would find out that sudo is not required.

simzacks commented on killFlask.sh:
```diff
@@ -0,0 +1,6 @@
+#FLASK_APP_FILE="flask_init.py"
+#kill -9 `ps aux | grep $FLASK_APP_FILE | grep -v grep | awk '{ print $2 }'`
+
+kill -9 `ps aux | grep flask | grep -v grep | awk '{ print $2 }'`
```
* get rid of unnecessary comments
* you need a new line at the end of the file.
* this function is a nice extra.

simzacks commented on runFlask.sh:
```diff
@@ -0,0 +1,14 @@
+
+VAG_COMMON_DIR="/vagrant"
+HOME="/home/vagrant"
+LOG_DIR="$VAG_COMMON_DIR/log"
```
i don't see where LOG_DIR is used here

DudiChen replied:
Just in the OUTFILE path, a line underneath it.

simzacks replied:
You may have intended this to be in the next file. In this file, I don't see neither an OUTFILE path nor its use.

DudiChen replied:
Sorry, misunderstanding on my part.
I initially thought of redirecting to a log file, but instead I referred it to /dev/null.
I will redirect it to a dedicated application log

simzacks commented on runFlask.sh:
```diff
@@ -0,0 +1,14 @@
+if [ -e venv/bin/activate ]; then
+    source venv/bin/activate
+    export FLASK_APP=$FLASK_APP_PY 
+    flask run --host=0.0.0.0 >> /dev/null 2>&1 &
```
Instead of logging here you send all output and errors to /dev/null

DudiChen replied:
I intended the OUTFILE to be the installation output of the setup.sh provision script, not as much as an application log for the Flask server.
I could add another log file for the Flask server.

simzacks replied:
There is no OUTFILE defined in this file

DudiChen replied:
Again, misunderstanding on my part.
As mentioned in previous comment, I will add a dedicated log for the Flask server.

simzacks commented on templates/index.html:
```diff
@@ -0,0 +1,10 @@
+    <img style="max-width: 100%" src="{{ url_for('static', filename='hello-world-image.png') }}">
+
+    </body>
+</html>
```
it is good practice to always have a newline at the end of a file

DudiChen replied:
Will do in next commit.

simzacks commented on setup.sh:
```diff
@@ -1,56 +1,64 @@
-    /usr/bin/python3 hello.py
+    echo "$((step++)). Initiating Flask Server..."
+    export FLASK_APP=$FLASK_APP_PY 
+    flask run --host=0.0.0.0 >> /dev/null 2>&1 &
```
Is there a reason you are throwing away all output and errors?

simzacks commented on runFlask.sh:
```diff
@@ -0,0 +1,14 @@
+HOME="/home/vagrant"
+LOG_DIR="$VAG_COMMON_DIR/log"
+FLASK_APP_PY="$VAG_COMMON_DIR/flask_init.py"
+FLASK_DIR=$HOME/FlaskApp
```
what's the purpose of this Dir? I don't see any files being put into it or used.

simzacks commented on setup.sh:
```diff
@@ -1,56 +1,64 @@
 LOG_DIR="$VAG_COMMON_DIR/log"
 OUTFILE="$LOG_DIR/${time}_vagrant-out.log"
+FLASK_APP_PY="$VAG_COMMON_DIR/flask_init.py"
+FLASK_DIR="$HOME/FlaskApp"
```
What's the purpose of this directory? I see you are creating it further down, but I don't see it being used at all.

DudiChen replied:
The script changes directory to this dir prior to setting the virtual environment, and running Flask.
all related virtual environment files are located in this dir.

DudiChen replied:
Will be removed in next commit, as I removed python's virtual environments.

simzacks commented on setup.sh:
```diff
@@ -1,56 +1,64 @@
+FLASK_DIR="$HOME/FlaskApp"
+FLASK_EXEC="$HOME/runFlask.sh"
+KILLFLASK_SCRIPT="$VAG_COMMON_DIR/killFlask.sh"
+PYTHON3="`which python3`"
```
executing "python3" will always give you the value returned by "which python3".
If the only think you are using this for is to execute python, just use the python3 executable.

DudiChen replied:
Noted. Will be applied in next commit.

simzacks commented on setup.sh:
```diff
@@ -1,56 +1,64 @@
     step=1
 
+    if [ ! -e $FLASK_APP_PY ]; then
+        echo "ERROR: Missing $FLASK_APP_PY file..." >> $OUTFILE
```
you are writing to OUTFILE here, which is in the LOG_DIR
A few lines down you create LOG_DIR if it doesn't exist.

DudiChen replied:
Fix will be applied in next commit.

simzacks commented on setup.sh:
```diff
@@ -1,56 +1,64 @@
     if [ ! -d $LOG_DIR ]; then
         mkdir $LOG_DIR
+    else
+        rm -f $LOG_DIR/*vagrant-out.log
```
Do you really want to delete all previous log files?

DudiChen replied:
I wanted to avoid accumulation of logs and requirement of maintenance.
For some retention, I guess I could delete all previous to the last 4 logs for example.

simzacks replied:
For application logs, you'll want to keep all of them with some sort of retention policy, which is probably out of scope for this. For installation logs, I would make sure to keep a few.
I would also suggest having a clean up script to deal with that and not build a retention policy inside the setup code.

I also don't think that we are up to that yet.

DudiChen replied:
Understood, so in order to keep the effort relevant I could redirect application log to a single log file in the meantime. And as mentioned will keep previous few for installation logs.

Should it be required, I could add a script for log retention (by size and number of previous logs) to be executed periodically via crontab.

simzacks commented on setup.sh:
```diff
@@ -1,56 +1,64 @@
+    echo "$((step++)). Installing & Activating Python-3 Virtual Environment..."
+    sudo apt-get install -yq python3-venv >> $OUTFILE 2>&1 
+    cd $FLASK_DIR
+    $PYTHON3 -m venv venv
```
why are you using a virtualenv?

DudiChen replied:
As I replied to Liora:

In short, a best practice to isolate a Python project with its own required libraries to ensure all dependencies.
It's reasonable that this won't necessarily be an issue in our case, I just applied it per Flask documentation recommendation.

Taken from Flask Documentation
https://flask.palletsprojects.com/en/1.1.x/installation/

"The more Python projects you have, the more likely it is that you need to work with different versions of Python libraries, or even Python itself. Newer versions of libraries for one project can break compatibility in another project.

Virtual environments are independent groups of Python libraries, one for each project. Packages installed for one project will not affect other projects or the operating system’s packages."

simzacks replied:
In general, its a very good idea. However, its not necessary when using vagrant if you are planning on using a single virtual machine per python project. It doesn't do any real harm, though, aside from adding complexity.

DudiChen replied:
Let me know should you prefer I remove it and perform tests accordingly.

simzacks replied:
either way works. Personally I wouldn't, but if the group wants to keep it as a standard best practice, I'd let it in

DudiChen replied:
Could be a good point to raise for best practice. 
However, I believe the team will appreciate simplicity, and truth be told I already removed and tested it :) so will be omitted in next commit,

simzacks reviewed (changes requested):
- the vagrantfile is much cleaner
- flask_init.py is better than having the code in a text file,reading and executing
- Using a file for html is good practice (as you did). For more complicated pages you will find that each page is a template. Often people try to build the whole page in the code, which gets very messy and unreadable.

DudiChen committed:
1. [setup.sh] Removed python3 venv installation and related directory.
2. [setup.sh] Removed python's path PYTHON variable.
3. [setup.sh] Moved $FLASK_APP_PY existance check past $OUTFILE.
4. [setup.sh] Replaced deletion of all logs to older than 5 most recent.
5. [setup.sh] Added redirection of Flask execution to a dedicated file.
   creation.
6. [runFlask.sh] Removed venv from Flask execution.
7. [runFlask.sh] Added redirection of Flask execution to a dedicated file.
8. [killFlask.sh] Removed comments.

simzacks commented on killFlask.sh:
```diff
@@ -1,6 +1,3 @@
-#FLASK_APP_FILE="flask_init.py"
-#kill -9 `ps aux | grep $FLASK_APP_FILE | grep -v grep | awk '{ print $2 }'`
-
 kill -9 `ps aux | grep flask | grep -v grep | awk '{ print $2 }'`
```
still need a new line at the end of the file

DudiChen replied:
I could swear I put it there. Will bee added next PR.
BTW, Indeed all code tested, and sudo never prompted for password.

simzacks commented on runFlask.sh:
```diff
@@ -1,14 +1,9 @@
-    flask run --host=0.0.0.0 >> /dev/null 2>&1 &
-fi
+export FLASK_APP=$FLASK_APP_PY 
+flask run --host=0.0.0.0 >> $FLASK_LOG 2>&1 &
```
new line at the end of file

DudiChen replied:
I could swear I put it there. Will bee added next PR.
BTW, Indeed all code tested, and sudo never prompted for password.

simzacks reviewed (changes requested):
Looking good. Have you tested all the code to make sure it works?
I don't see anywhere that your sudo is setup as not needing a password.
Is that automatic?

lmilbaum reviewed (approved):
Approving this PR to unblock the process. Please address all technical debts in the next PRs.
